modules/grandpa/src/lib.rs

```diff
@@ -248,6 +265,15 @@ pub mod pallet {
 	#[pallet::storage]
 	pub(super) type BestFinalized<T: Config<I>, I: 'static = ()> = StorageValue<_, BridgedBlockHash<T, I>, ValueQuery>;
 
+	/// A ring buffer of imported hashes. Ordered by the insertion time.
+	#[pallet::storage]
+	pub(super) type ImportedHashes<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Identity /* TODO */, u32, BridgedBlockHash<T, I>>;
+
+	/// Current ring buffer position.
+	#[pallet::storage]
+	pub(super) type ImportedHashesNeedle<T: Config<I>, I: 'static = ()> = StorageValue<_, u32, ValueQuery>;
```

Comment:
I would prefer to use the term `index`, `position` or `pointer` here. `Needle`, AFAIK, is not commonly used.